src/Microsoft.OData.Core/UriParser/Parsers/FunctionOverloadResolver.cs

```diff
@@ -164,45 +165,38 @@ internal static bool ResolveOperationFromList(string identifier, IEnumerable<str
                 throw;
             }
 
-            IList<IEdmOperation> foundActionsWhenLookingForFunctions = new List<IEdmOperation>();
-            bool hasParameters = parameterNames.Count() > 0;
+            IList<IEdmOperation> foundActionsWhenLookingForFunctions = null;
+            bool hasParameters = parameterNames.Count > 0;
 
-            if (bindingType != null)
-            {
-                candidateMatchingOperations.EnsureOperationsBoundWithBindingParameter();
-            }
+            IList<IEdmOperation> candidatesMatchingOperations = operationsFromModel.FilterOperationCandidatesBasedOnParameterList(bindingType, parameterNames, resolver.EnableCaseInsensitive, out foundActionsWhenLookingForFunctions);
 
-            // If the number of parameters > 0 then this has to be a function as actions can't have parameters on the uri only in the payload. Filter further by parameters in this case, otherwise don't.
-            if (hasParameters)
-            {
-                // can only be a function as only functions have parameters on the uri.
-                candidateMatchingOperations = candidateMatchingOperations.RemoveActions(out foundActionsWhenLookingForFunctions)
-                    .FilterOperationsByParameterNames(parameterNames, resolver.EnableCaseInsensitive);
-            }
-            else if (bindingType != null)
-            {
-                // Filter out functions with more than one parameter. Actions should not be filtered as the parameters are in the payload not the uri
-                candidateMatchingOperations = candidateMatchingOperations.Where(o =>
-                (o.IsFunction() && (o.Parameters.Count() == 1 || o.Parameters.Skip(1).All(p => p is IEdmOptionalParameter))) || o.IsAction());
-            }
-            else
+            // Only filter if there is more than one and its needed.
+            if (candidatesMatchingOperations.Count > 1)
             {
-                // Filter out functions with any parameters
-                candidateMatchingOperations = candidateMatchingOperations.Where(o => (o.IsFunction() && !o.Parameters.Any()) || o.IsAction());
+                candidatesMatchingOperations = candidatesMatchingOperations.FilterBoundOperationsWithSameTypeHierarchyToTypeClosestToBindingType(bindingType) as IList<IEdmOperation>;
             }
 
-            // Only filter if there is more than one and its needed.
-            if (candidateMatchingOperations.Count() > 1)
+            // If any of the things returned are an action, it better be the only thing returned, and there can't be parameters in the URL
+            bool actionExists = false;
+            bool functionExists = false;
+            for (int i = 0; i < candidatesMatchingOperations.Count; i++)

```

Comment:
>for [](start = 12, length = 3)

Should we break the loop as soon as functionExists and/or actionExists ?

Reply:
We can break when both exists. 

---
In reply to: [354092013](https://github.com/OData/odata.net/pull/1602#discussion_r354092013) [](ancestors = 354092013)